File: src/HHLookupTables.cpp
#include "HHLookupTables.h"
#include <cmath>
#include <vector>

// --- Private kinetic functions, used only for initialization ---

double HHLookupTables::alpha_m(double V) {
    // A small offset to avoid division by zero at V = -40
    if (std::abs(V + 40.0) < 1e-5) {
        return 1.0;
    }
    return 0.1 * (V + 40.0) / (1.0 - exp(-(V + 40.0) / 10.0));
}

double HHLookupTables::beta_m(double V) {
    return 4.0 * exp(-(V + 65.0) / 18.0);
}

double HHLookupTables::alpha_h(double V) {
    return 0.07 * exp(-(V + 65.0) / 20.0);
}

double HHLookupTables::beta_h(double V) {
    return 1.0 / (1.0 + exp(-(V + 35.0) / 10.0));
}

double HHLookupTables::alpha_n(double V) {
    // A small offset to avoid division by zero at V = -55
    if (std::abs(V + 55.0) < 1e-5) {
        return 0.1;
    }
    return 0.01 * (V + 55.0) / (1.0 - exp(-(V + 55.0) / 10.0));
}

double HHLookupTables::beta_n(double V) {
    return 0.125 * exp(-(V + 65.0) / 80.0);
}

// --- Public static method to initialize the tables ---

void HHLookupTables::initialize() {
    // Resize vectors to the correct size
    alpha_m_lut.resize(LUT_SIZE);
    beta_m_lut.resize(LUT_SIZE);
    alpha_h_lut.resize(LUT_SIZE);
    beta_h_lut.resize(LUT_SIZE);
    alpha_n_lut.resize(LUT_SIZE);
    beta_n_lut.resize(LUT_SIZE);

    // Populate the tables
    for (int i = 0; i < LUT_SIZE; ++i) {
        double v = V_MIN + i * V_STEP;
        alpha_m_lut[i] = alpha_m(v);
        beta_m_lut[i]  = beta_m(v);
        alpha_h_lut[i] = alpha_h(v);
        beta_h_lut[i]  = beta_h(v);
        alpha_n_lut[i] = alpha_n(v);
        beta_n_lut[i]  = beta_n(v);
    }
}

File: src/gpu_simulation/hh_lookup_tables.py
import torch
import numpy as np

class HHLookupTables:
    V_MIN = -100.0
    V_MAX = 50.0
    V_STEP = 0.01
    LUT_SIZE = int((V_MAX - V_MIN) / V_STEP) + 1

    def __init__(self, device='cpu', dtype=torch.float64):
        self.device = device
        self.dtype = dtype
        self.alpha_m_lut = None
        self.beta_m_lut = None
        self.alpha_h_lut = None
        self.beta_h_lut = None
        self.alpha_n_lut = None
        self.beta_n_lut = None
        self.initialize()

    def _alpha_m(self, V):
        V_shifted = V + 40.0
        # Use torch.where to handle the singularity
        return torch.where(torch.abs(V_shifted) < 1e-5,
                           torch.tensor(1.0, device=self.device),
                           0.1 * V_shifted / (1.0 - torch.exp(-V_shifted / 10.0)))

    def _beta_m(self, V):
        return 4.0 * torch.exp(-(V + 65.0) / 18.0)

    def _alpha_h(self, V):
        return 0.07 * torch.exp(-(V + 65.0) / 20.0)

    def _beta_h(self, V):
        return 1.0 / (1.0 + torch.exp(-(V + 35.0) / 10.0))

    def _alpha_n(self, V):
        V_shifted = V + 55.0
        # Use torch.where to handle the singularity
        return torch.where(torch.abs(V_shifted) < 1e-5,
                           torch.tensor(0.1, device=self.device),
                           0.01 * V_shifted / (1.0 - torch.exp(-V_shifted / 10.0)))

    def _beta_n(self, V):
        return 0.125 * torch.exp(-(V + 65.0) / 80.0)

    def initialize(self):
        v_range = torch.linspace(self.V_MIN, self.V_MAX, self.LUT_SIZE, device=self.device, dtype=self.dtype)

        self.alpha_m_lut = self._alpha_m(v_range)
        self.beta_m_lut = self._beta_m(v_range)
        self.alpha_h_lut = self._alpha_h(v_range)
        self.beta_h_lut = self._beta_h(v_range)
        self.alpha_n_lut = self._alpha_n(v_range)
        self.beta_n_lut = self._beta_n(v_range)

if __name__ == '__main__':
    # Example of how to use it
    luts = HHLookupTables(device='cuda' if torch.cuda.is_available() else 'cpu')
    print(f"LUTs created on device: {luts.device}")
    print(f"LUT size: {luts.LUT_SIZE}")
    print(f"Alpha_m LUT shape: {luts.alpha_m_lut.shape}")
    # Print first 5 values of alpha_m
    print("First 5 alpha_m values:")
    print(luts.alpha_m_lut[:5])
